File: src/advertising_monitor.py
import csv
import json
import time
import os
import subprocess


def read_csv(file_path):
    data = []
    with open(file_path, mode='r') as csv_file:
        csv_reader = csv.DictReader(csv_file)
        print("Reading CSV file...")
        for row in csv_reader:
            data.append(row)
    print(f"Read {len(data)} records from CSV.")
    return data


def advertise(uuid, is_valid):
    manufacturer_data = f"{uuid}{'01' if is_valid else '00'}"
    manufacturer_data = manufacturer_data.encode('utf-8').hex()
    duration = 300000  # 300 seconds in milliseconds
    command = ['node', 'advertise.js', manufacturer_data, str(duration)]
    print(f"Running advertising command for UUID {uuid}...")
    subprocess.run(command, cwd=os.path.dirname(__file__))
    print(f"Advertising command completed for UUID {uuid}.")


if __name__ == "__main__":
    base_dir = os.path.dirname(os.path.dirname(
        __file__))  # Go up one level from src
    input_file_path = os.path.join(base_dir, 'processed_tickets.csv')
    last_seen_data = set()

    print("Starting main loop of the ticket processing and advertising system...")
    while True:
        if os.path.exists(input_file_path):
            print(f"Checking for new data in {input_file_path}")
            current_data = set(tuple(d.items())
                               for d in read_csv(input_file_path))
            new_entries = current_data - last_seen_data
            if new_entries:
                print(
                    f"Found {len(new_entries)} new entries. Processing new entries...")
            for entry in new_entries:
                data = dict(entry)
                uuid = data['UUID']
                is_valid = data['is_valid'] == 'True'
                print(f"Advertising for UUID {uuid} with validity {is_valid}")
                advertise(uuid, is_valid)
            last_seen_data = current_data
        else:
            print(
                f"No file found at {input_file_path}, will check again in 10 seconds.")
        time.sleep(10)

File: src/advertising_monitor.js
const fs = require('fs');
const path = require('path');
const csv = require('csv-parser');
const { exec } = require('child_process');

const baseDir = path.dirname(path.dirname(__filename));
const inputFilePath = path.join(baseDir, 'fake_processed_tickets.csv');
let lastSeenData = new Set();

function readCsv(filePath) {
    return new Promise((resolve, reject) => {
        const data = [];
        fs.createReadStream(filePath)
            .pipe(csv())
            .on('data', (row) => data.push(row))
            .on('end', () => {
                console.log(`Read ${data.length} records from CSV.`);
                resolve(data);
            })
            .on('error', reject);
    });
}

function advertise(uuid, isValid) {
    return new Promise((resolve, reject) => {
        const manufacturerData = `${uuid}${isValid ? '01' : '00'}`;
        const duration = 300000; // 300 seconds in milliseconds
        const command = `node advertise.js ${manufacturerData} ${duration}`;
        console.log(`Running advertising command for UUID ${uuid}...`);

        exec(command, { cwd: __dirname }, (error, stdout, stderr) => {
            if (error) {
                console.error(`Error advertising UUID ${uuid}: ${error}`);
                return reject(error);
            }
            console.log(`Advertising command completed for UUID ${uuid}.`);
            console.log(stdout);
            console.error(stderr);
            resolve();
        });
    });
}

async function main() {
    console.log("Starting main loop of the ticket processing and advertising system...");
    while (true) {
        if (fs.existsSync(inputFilePath)) {
            console.log(`Checking for new data in ${inputFilePath}`);
            const currentData = new Set((await readCsv(inputFilePath)).map(JSON.stringify));
            const newEntries = [...currentData].filter(entry => !lastSeenData.has(entry));
            if (newEntries.length > 0) {
                console.log(`Found ${newEntries.length} new entries. Processing new entries...`);
                for (const entry of newEntries) {
                    const data = JSON.parse(entry);
                    const uuid = data['UUID'];
                    const isValid = data['is_valid'] === 'True';
                    console.log(`Advertising for UUID ${uuid} with validity ${isValid}`);
                    await advertise(uuid, isValid);
                }
                lastSeenData = currentData;
            }
        } else {
            console.log(`No file found at ${inputFilePath}, will check again in 10 seconds.`);
        }
        await new Promise(resolve => setTimeout(resolve, 10000));
    }
}

main().catch(console.error);
